### ISSUE TITLE
ParallelAssignment false positive

### ISSUE THREAD

**Issue by Zmote** [rubocop-hq/rubocop#8217]
The autofix for ParallelAssignment leads to an error in logic.

## Expected behavior

```ruby
@item.position, @item2.position = @item2.position, @item.position
```

Should resolve to:

```ruby
original_item_position = @item.position
@item.position = @item2.position
@item2.position = original_item_position
```

The autofix should introduce a temporary variable here or another approach, where the original value isn't overwritten.

## Actual behavior

Currently, the above resolves to:

```ruby
@item.position = @item2.position
@item2.position = @item.position
```

## Steps to reproduce the problem

An important context to consider for reproduction: @item and @item2 are active record instances, position is a numeric value.

## RuboCop version

```
$ [bundle exec] rubocop -V
0.86.0 (using Parser 2.7.1.4, rubocop-ast 0.1.0, running on jruby 2.5.7 java)
```

Might also be interesting, additional gems I have linked in the contet of rubocop:

```ruby
  gem 'rubocop', '0.86.0', require: false
  gem 'rubocop-rspec', '1.40.0', require: false
  gem 'rubocop-performance', '1.6.1', require: false
  gem 'rubocop-rails', '2.6.0', require: false
```

**Comment by marcandre** [comment]
Agree with the bug report, except the cop should simply accept it, as it already accepts `a, b = b, a`

**Comment by Zmote** [comment]
Interesting point, that is true, that with a simple parallel assignment statement (default settings), no violation is shown, only in this object context. In that case I think it would make sense to extend the bug report so that not only the autofix needs to be adjusted, but the cop as well.

**Comment by marcandre** [comment]
Yes (except that the autofix doesn't need to be adjusted, just the cop). I've taken the liberty of editing your title.

**Comment by Zmote** [comment]
How so? Could you elaborate? Let's say in my rubocop YAML - File I decide to disallow any kind of Parallel Assignment. The autofix provided doesn't "fix" it to correct form and leads to a bug in code. I have a certain location in code where this is used to swap page ordering and due to this, sooner or later, all pages move towards one and the same position number, which doesn't happen in the "before fix" form, ie. parallel assignment. The fix leads to a code, where instead of a swap, an overwrite occurs. What am I missing?

**Comment by marcandre** [comment]
> Let's say in my rubocop YAML - File I decide to disallow any kind of Parallel Assignment

Unless I'm mistaken, there is no such setting. I don't think there should be one either; such parallel assignment is both convenient and expressive.

**Comment by Zmote** [comment]
Hmm, O.K., I see how the cop works then. In that case fixing the cop should suffice. Thanks for the clarification.

Whether or not there should be an option to disallow parallel assignment or not: I think it would be interesting to have it, if one might deem use of parallel assignment not convenient / not expressive for their code base, but I assume that's too much of an edge case to provide as an option.

**Comment by rrosenblum** [comment]
Hmm, this issue sounds familiar. This might be an edge case of an existing condition that we check for. This seems related to this commit `00b4c9034 2016-10-01 17:39:31 +0200 Jonas Arvidsson | [Fix #3286] Handle `self.a, self.b = b, a` in ParallelAssignment`. I agree with @marcandre interpretation that this cop should be modified to allow the case of variable reassignment (variable swapping?).

**Comment by nguyenquangminh0711** [comment]
Am I misunderstand something, but the current implementation of Style/ParallelAssignment already accepts the variable reassignment at https://github.com/rubocop-hq/rubocop/blob/master/lib/rubocop/cop/style/parallel_assignment.rb#L56 by using topolocial sort. It's also stated in the [test](https://github.com/rubocop-hq/rubocop/blob/master/spec/rubocop/cop/style/parallel_assignment_spec.rb#L241). I tested the sample code but found no offenses in the latest rubocop version.

**Comment by fatkodima** [comment]
@marcandre As a previous commenter, I can't reproduce this on a stated version (`0.86`) nor on `master`. 
I think, this should be closed.
